### PR TITLE
Gluon trainer updates: add learning_rate and lr_scheduler properties and add setter for learning rate

### PR DESCRIPTION
**Pull request by astonzhang**

### CODE REVIEW

File: python/mxnet/gluon/trainer.py [hunk 1]
```diff
@@ -113,6 +113,36 @@ def _init_kvstore(self):
 
         self._kv_initialized = True
 
+
+    @property
+    def learning_rate(self):
```
Comment (piiswrong) [inline python/mxnet/gluon/trainer.py]: document this as 
```Properties
----------```

in init doc

File: python/mxnet/gluon/trainer.py [hunk 2]
```diff
@@ -113,6 +113,36 @@ def _init_kvstore(self):
...
+        return self._optimizer.lr
+
+
+    @property
+    def lr_scheduler(self):
```
Comment (piiswrong) [inline python/mxnet/gluon/trainer.py]: don't expose this for now

File: python/mxnet/gluon/trainer.py (hunk 1 above)
Comment (piiswrong) [reply python/mxnet/gluon/trainer.py]: Also report learning_rate when using lr_sheduler

File: python/mxnet/gluon/trainer.py [hunk 3]
```diff
@@ -113,6 +113,36 @@ def _init_kvstore(self):
 
         self._kv_initialized = True
 
+
+    @property
+    def learning_rate(self):
+        return self._optimizer.lr
```
Comment (szha) [inline python/mxnet/gluon/trainer.py]: This kind of implementation couples the two classes together (i.e. Trainer must know the structure of optimizer). Instead add accessors in optimizer and make reading learning rate the concern of optimizer.

File: python/mxnet/gluon/trainer.py [hunk 4]
```diff
@@ -113,6 +113,36 @@ def _init_kvstore(self):
...
+        """
+        if not isinstance(self._optimizer, opt.Optimizer):
+            raise UserWarning("Optimizer has to be defined before its learning"
+                              "rate is mutated.")
+        elif self._optimizer.lr_scheduler is not None:
+            raise UserWarning("set_learning_rate mutates the value of the"
+                              "learning rate only when the LRScheduler of"
+                              "the optimizer is undefined.")
+        else:
+            self._optimizer.lr = lr
```
Comment (szha) [inline python/mxnet/gluon/trainer.py]: Same as the accessor comment. Try make setting learning rate the concern of optimizer's.

File: python/mxnet/gluon/trainer.py (hunk 1 above)
Comment (astonzhang) [reply python/mxnet/gluon/trainer.py]: resolved

File: python/mxnet/gluon/trainer.py (hunk 3 above)
Comment (astonzhang) [reply python/mxnet/gluon/trainer.py]: resolved

File: python/mxnet/gluon/trainer.py (hunk 4 above)
Comment (astonzhang) [reply python/mxnet/gluon/trainer.py]: resolved

File: python/mxnet/gluon/trainer.py (hunk 2 above)
Comment (astonzhang) [reply python/mxnet/gluon/trainer.py]: resolved

File: python/mxnet/optimizer.py [hunk 5]
```diff
@@ -191,6 +204,24 @@ def update(self, index, weight, grad, state):
         """
         raise NotImplementedError()
 
+    def set_learning_rate(self, lr):
```
Comment (szha) [inline python/mxnet/optimizer.py]:
```
@learning_rate.setter
def set_learning_rate
```
and then you can do
```
optimizer.learning_rate = 0.5
```

File: python/mxnet/gluon/trainer.py [hunk 6]
```diff
@@ -113,6 +119,34 @@ def _init_kvstore(self):
...
+        Parameters
+        ----------
+        lr : float
+            The new learning rate of the optimizer.
+        """
+        if not isinstance(self._optimizer, opt.Optimizer):
+            raise UserWarning("Optimizer has to be defined before its learning"
+                              "rate is mutated.")
+        else:
+            self._optimizer.set_learning_rate(lr)
```
Comment (szha) [inline python/mxnet/gluon/trainer.py]: Update according to the comment below

File: python/mxnet/optimizer.py (hunk 5 above)
Comment (astonzhang) [reply python/mxnet/optimizer.py]: resolved

File: python/mxnet/gluon/trainer.py (hunk 6 above)
Comment (astonzhang) [reply python/mxnet/gluon/trainer.py]: resolved

File: python/mxnet/optimizer.py (hunk 5 above)
Comment (astonzhang) [reply python/mxnet/optimizer.py]: resolved

File: python/mxnet/gluon/trainer.py [hunk 7]
```diff
@@ -43,6 +43,12 @@ class Trainer(object):
     kvstore : str or KVStore
         kvstore type for multi-gpu and distributed training. See help on
         :any:`mxnet.kvstore.create` for more information.
+
+    Properties
+    ----------
+    Learning_rate: float
+        The learning rate of the optimizer or the learning rate of the
+        LRScheduler of the optimizer if the LRScheduler is defined.
```
Comment (piiswrong) [inline python/mxnet/gluon/trainer.py]: explain it can be accessed with trainer.learning_rate and set with trainer.learning_rate = xxx

Comment (astonzhang) [reply python/mxnet/gluon/trainer.py]: resolved

**Comment by szha** [comment]
All tests passed except R GPU, @thirdwing is fixing this in #7686.

File: tests/python/unittest/test_loss.py [hunk 8]
```diff
@@ -70,15 +70,14 @@ def test_ce_loss():
     label = mx.nd.array(np.random.randint(0, nclass, size=(N,)), dtype='int32')
```
Comment (piiswrong) [inline tests/python/unittest/test_loss.py]: Submit test fix in a separate PR and revert the change on file property
100644 → 100755

Comment (astonzhang) [reply tests/python/unittest/test_loss.py]: resolved

**Comment by astonzhang** [comment]
The unit test patch is in a separate pr as suggested by @piiswrong 
#7693

**Comment by astonzhang** [comment]
Closing this PR due to new PR at #7760
